components/timezone/src/types.rs

```diff
@@ -35,6 +44,85 @@ impl GmtOffset {
         Self(0)
     }
 
+    /// Parse a [`GmtOffset`] from bytes.
+    ///
+    /// The offset must range from GMT-12 to GMT+14.
+    /// The string must be an ISO-8601 time zone designator:
+    /// e.g. Z
+    /// e.g. +05
+    /// e.g. +0500
+    /// e.g. +05:00
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use icu::timezone::GmtOffset;
+    /// use icu::timezone::TimeZoneError;
+    ///
+    /// let offset0: GmtOffset =
+    ///     GmtOffset::try_from_bytes(b"Z").expect("Failed to parse a time zone");
+    /// let offset1: GmtOffset =
+    ///     GmtOffset::try_from_bytes(b"+05").expect("Failed to parse a time zone");
+    /// let offset2: GmtOffset =
+    ///     GmtOffset::try_from_bytes(b"+0500").expect("Failed to parse a time zone");
+    /// let offset3: GmtOffset =
+    ///     GmtOffset::try_from_bytes(b"-05:00").expect("Failed to parse a time zone");
+    /// let offset_err: TimeZoneError =
+    ///     GmtOffset::try_from_bytes(b"0500").expect_err("Invalid input");
+    ///
+    /// assert_eq!(offset0.offset_seconds(), 0);
+    /// assert_eq!(offset1.offset_seconds(), 18000);
+    /// assert_eq!(offset2.offset_seconds(), 18000);
+    /// assert_eq!(offset3.offset_seconds(), -18000);
+    /// assert_eq!(offset_err, TimeZoneError::InvalidOffset);
+    /// ```
+    pub fn try_from_bytes(chars: &[u8]) -> Result<Self, TimeZoneError> {
+        let offset_sign = match chars.first() {
+            Some(b'+') => 1,
+            Some(b'-') => -1,
+            Some(b'Z') => return Ok(Self(0)),
+            _ => return Err(TimeZoneError::InvalidOffset),
+        };
+
+        let rest = &chars[1..];
```

Comment:
issue: we're trying to avoid any panicky APIs. It's clear here that this won't panic, but it's easier to just not deal with the possibility. Can you please change it back to using iterators or some other way to avoid indexing (like `split_at()`.)?